### DISCUSSION TITLE
Interaction with Microsoft.NET.Test.Sdk

### DISCUSSION THREAD

**Discussion by seanswallingcc** [xunit/xunit#2787] [Question] [answered]
Hi. 

I'm currently involved in a large scale upgrade of systems from legacy projects in .net framework towards sdk projects and .net6+.
As such our current state is that we use xunit1.9.2 with an often arbitrary version for xunit.runner.visualstudio currently at .NetFramework 4.8

As a part of this upgrade I've ended up doing two things:
- unifying xunit.runner.visualstudio to 2.5.1 in our test projects
- importing Microsoft.Net.Test.Sdk 17.4.1

However, some of our tests and infrastructure end up running a scan across our assemblies and run a `assembly.GetTypes()` on xunit.runner.visualstudio which throws the following exception:

> Message: Could not load file or assembly 'Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

I have been doing some research in this space, but I am still unsure as to the resolution.
- Microsoft.VisualStudio.TestPlatform.ObjectMode is not listed as a dependency for Microsoft.Net.Test.Sdk in NETFramework 4.6.2,  adding it as a dependency (even though it is unused) would "solve" the problem although it leave me unsure if I am doing the correct thing from the pov of xunit and Microsoft.Net.Test.Sdk.
- I could also just modify the code that does the scanning to exclude xunit.runner.visualstudio, but also this leaves me unsure if I am doing the correct thing.  This code used to run fine previously, so presumably it should continue to do so unless there was a significant change the in the VS runner.

It's also possible I have my version numbers wrong, I am aware that xunit has decoupled their version dependency from Microsoft.Net.Test.Sdk, but I do wonder if I stumbled on an incompatibility?  I couldn't find any documentation of version compatibilities.

I'd appreciate any insight into this matter, in case I have done something obviously incorrect, or misunderstood the relationship between XUnit and .NET Test SDK.  Thanks.

**Comment by bradwilson** [comment, marked as answer]
Some assemblies are provided dynamically at runtime by the VSTest environment, including `Microsoft.VisualStudio.TestPlatform.ObjectModel`. You should not need this as a reference. The reason it does not show up as a dependency is because the `xunit.runner.visualstudio` NuGet package isn't a "library package"; that is, by importing it you're not getting something linked into your code. If you look at the `lib` folder inside of it, you'll see that there are no libraries for you to link against; instead, the `build` folder contains a `.props` file that gets injected into the MSBuild pipeline for any project which references the package (in this case, copying the binaries to output folder so VSTest can find them).

Trying to load the test adapter dynamically outside of VSTest isn't a supported scenario. That said, [`Assembly.GetTypes()`](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.gettypes) is absolutely shown to throw exceptions, so you should be catching [`ReflectionTypeLoadException`](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.reflectiontypeloadexception) and using the `.Types` property on it to get the types that can be provided. This actually has nothing to do with us; it's just generally how `GetTypes` should be used.

I think hard-coding a skip of our test adapter should be fine, especially if it's not going to provide any value to your scanning process.

**Reply by seanswallingcc** [reply]
Hi Brad.

Thanks for you answer.

For clarity, we scan all referenced assemblies to build a dictionary of sorts for Production, and in unit test scenarios we swap this for scanning all assemblies in the app domain. The unfortunate situation happens because the runner gets returned from the list of assemblies in the app domain, so our production logic scans it like normal, before failing due to above.

Happy to exclude the runner from this logic, as ideally we wouldn't bother scanning the runner.
